### DISCUSSION TITLE
GitHub Copilot High Resource Usage Bug Report

### DISCUSSION THREAD

**Discussion by koen-dev** [community/community#56755] [Copilot Conversations]
### Description

After upgrading the GitHub Copilot extension to the version 1.82.51 in VS Code, the `rg` process is experiencing high CPU usage, specifically running at over 500% on an Apple M1 MacBook Pro. This issue becomes particularly apparent when working on larger projects, rendering the GitHub Copilot extension essentially unusable. It appears that this is not an isolated incident, as other users have reported similar experiences. Here is a link to a previous discussion on this topic: [Previous Discussion Comment](https://github.com/orgs/community/discussions/53846#discussioncomment-5770181).

### Steps to Reproduce

1. Upgrade the GitHub Copilot extension to version 1.82.51 (or later) in VS Code.
2. Open a larger project in VS Code with the GitHub Copilot extension enabled.
3. Monitor the resource usage of the `rg` process.

### Expected Result

The `rg` process should use a reasonable amount of resources, allowing for smooth and efficient operation of the VS Code and the GitHub Copilot extension even on larger projects.

### Actual Result

The `rg` process is consuming excessive resources, specifically running at over 500% CPU usage. This leads to significant slowdown and makes it impossible to effectively use the GitHub Copilot extension, especially on larger projects.

Given the subscription model of the GitHub Copilot extension, the expectation is that this issue would be addressed, or at least acknowledged, by GitHub in a timely manner. Currently, that has not happened, which is a source of frustration.

**Comment by Akash1134** [comment]
Hi @koen-dev ,

Thanks for reporting this issue with such a great description. Apologies for the trouble caused.

I have shared this issue internally with the product team and they are looking into it. 

Please note that issues like these need thorough investigation and enough votes to be able to identify as a bug and it will need some time. We do not have an ETA on our response for the outcome. That being said, this is on our radar and if we notice the issue to be more prevalent, we'll do the needful. 🙂

**Reply by koen-dev** [reply]
Hi @Akash1134,

Thanks for jumping on this so quickly, I totally get that these things can be a bit like hunting for a needle in a code-stack! No worries about the timeline, it's comforting to know it's in good hands.

Really appreciate the acknowledgement and the effort to look into this. Looking forward to hearing about any updates even if it isn't what I hope for.

**Reply by rahulchauhan04** [reply]
woo h yaa .....

its been 7 days i am facing that issue....
now my system is working fine with ...

i have just added the regex to get easier for copilot to ignore this file... earlier i have used the copilotignore .gitignore but none of it works...

here is the vscode setting.json which you can use fi you want...

i am sharign here because i got some good insights from this chat to solve this probelm..

"github.copilot.advanced": {
        "ignoredFiles": {
            "**/node_modules/**": true,
            "**/dist/**": true,
            "**/build/**": true,
            "**/*.min.js": true
        },
        "ignoredFolders": {
            "**/node_modules": true,
            "**/dist": true,
            "**/build": true
        },
        "ignoredFilesGlob": [
            "**/node_modules/**",
            "**/dist/**",
            "**/build/**",
            "**/*.min.js"
        ],
        "ignoredFoldersGlob": [
            "**/node_modules",
            "**/dist",
            "**/build"
        ],
        "ignoredFilesRegex": [
            "node_modules",
            "dist",
            "build",
            ".min.js"
        ],
        "ignoredFoldersRegex": [
            "node_modules",
            "dist",
            "build"
        ],
        "ignoredFilesGlobRegex": [
            "node_modules",
            "dist",
            "build",
            ".min.js"
        ],
        "ignoredFoldersGlobRegex": [
            "node_modules",
            "dist",
            "build"
        ],
        "ignoredFilesGlobRegexArray": [
            "node_modules",
            "dist",
            "build",
            ".min.js"
        ],
        "ignoredFoldersGlobRegexArray": [
            "node_modules",
            "dist",
            "build"
        ],
        "ignoredFilesArray": [
            "**/node_modules/**",
            "**/dist/**",
            "**/build/**",
            "**/*.min.js"
        ],
        "ignoredFoldersArray": [
            "**/node_modules",
            "**/dist",
            "**/build"
        ],
        "ignoredFilesGlobArray": [
            "**/node_modules/**",
            "**/dist/**",
            "**/build/**",
            "**/*.min.js"
        ],
        "ignoredFoldersGlobArray": [
            "**/node_modules",
            "**/dist",
            "**/build"
        ],
        "ignoredFilesRegexArray": [
            "node_modules",
            "dist",
            "build",
            ".min.js"
        ],
        "ignoredFoldersRegexArray": [
            "node_modules",
            "dist",
            "build"
        ],
        
    },

**Reply by serhiiokhotnik** [reply]
my vscode says it's `Unknown Configuration Setting`

**Comment by airtonix** [comment]
`1.82.15` is the last version with normal cpu/gpu usage.

https://github.com/orgs/community/discussions/53846#discussioncomment-5922539

**Comment by iamobj** [comment]
High cpu usage, resulting in large cpu cooling fan speed, noise and computer slowdown, especially in multiple open vscode windows, the computer will become slow, because the cpu resources are occupied too much

I tried to use vscode remote development to free up my local computer, however Copilot is still taking up my local cpu resources

**Comment by marcossv9** [comment]
Having the same issue here with high cpu usage when copilot extension is enabled.. note that I'm running Ubuntu and **not** MacOS.

Here are the versions of my setup:
```
code --version
1.79.2
695af097c7bd098fbf017ce3ac85e09bbc5dda06
x64
---
cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.2 LTS"
---
Github Copilot version: v1.94.220
```
The only way I could stop the high cpu usage was disabling the copilot extension..

**Reply by marcossv9** [reply]
The way I ended up resolving this, was by adding this parameter to vscode settings:
`"search.followSymlinks": false`
Hope it helps.

**Reply by ashleybartlett** [reply]
This worked for my use case. Seems like a workaround. Might not be an option for everyone, as some repo's feature symlinks.

**Reply by hpolymath** [reply]
I just adde the copilot to VSCode and my MacBook Pro of 4 years has turned its fans for the first time ever. can you explain to me how to use this and what exactly does it do?

**Reply by matthew-heyner** [reply]
Run `ps aux | top` and kill the copilot extension using all your CPU... Every time my fans spin up it's 9 times out of 10 copilot...

**Reply by rahulchauhan04** [reply]
it is reading Large files,,,eg,, nodemodules.. analyzingn those files make the cpu usage spike alot

**Comment by jshtuazon2** [comment]
code --version
1.79.2
695af097c7bd098fbf017ce3ac85e09bbc5dda06
x64
---
cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.2 LTS"
---
Github Copilot version: v1.94.220

**Comment by jshtuazon2** [comment]
CODE_OF_CONDUCT.md

**Comment by pierre-cm** [comment]
I'm exepriencing same issue when running `github.copilot.generate` command.

When I try to log the command line of the process causing excessive memory and cpu usage i get the following:

```shell
$ cat /proc/79906/cmdline
/usr/share/code/code--ms-enable-electron-run-as-node--max-old-space-size=3072/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js--useInferredProjectPerProjectRoot--enableTelemetry--cancellationPipeName/tmp/vscode-typescript1000/018118fd9725797d1978/tscancellation-3f34ff507dbd39ee5a10.tmp*--localeen--noGetErrOnBackgroundUpdate--validateDefaultNpmLocation--useNodeIpc
```

It seems that it is somehow related to the way the extension is interracting with TypeScript Language Service.

```
copilot version
v1.94.220
code --version
1.79.2
695af097c7bd098fbf017ce3ac85e09bbc5dda06
x64
```

**Comment by ashleybartlett** [comment]
I've run into this problem today while getting setup on a new monorepo. 

This is the rg command consuming CPU on my laptop:
```shell
/Applications/Visual Studio Code.app/Contents/Resources/app/node_modules.asar.unpacked/@vscode/ripgrep/bin/rg --files --hidden --case-sensitive --no-require-git -g **/.copilotignore -g !**/.git -g !**/.svn -g !**/.hg -g !**/CVS -g !**/.DS_Store -g !**/Thumbs.db --no-ignore --follow --no-config --no-ignore-global
```

Deciphering this command, it believe it's looking for `.copilotignore` files, and is not respecting ignore options.

I ran the above command directly, and it looks like some of our node modules have circular dependencies, which are causing excessive recursion. Even though rg has some detection for this and stops before going around the world again, if it keeps on hitting this problem it would significantly extend the runtime of this command. 

copilot version: v1.99.289
vscode version: 1.80.1
74f6148eb9ea00507ec113ec51c489d6ffb4b771
x64

**Reply by ashleybartlett** [reply]
I did a quick test, and I ran the same command, excluding `**/node_modules/**` and the command finished almost instantly. I tried to be specific to one module, but there were more circular references than I had expected!

```shell
/Applications/Visual\ Studio\ Code.app/Contents/Resources/app/node_modules.asar.unpacked/@vscode/ripgrep/bin/rg --files --hidden --case-sensitive --no-require-git -g '**/.copilotignore' -g '*/.git' -g '!**/.svn' -g '!**/.hg' -g '!**/CVS' -g '!**/.DS_Store' -g '!**/Thumbs.db' --no-ignore --follow --no-config --no-ignore-global -g '!**/node_modules/**' 
```

**Reply by koen-dev** [reply]
This is definitely it! It looks like what ever you put in a `.copilotignore` it isn't correctly being passed on to the `rg` command.

@Akash1134 any updates from your side perhaps about the issue in general or the issue mentioned above?

**Reply by ashleybartlett** [reply]
ah, so it's not the `.copilotignore` existing itself, it's due to a large number of node_modules that have circular depedencies in a monorepo. From what I was able to determine, my inference is that this file lookup is using a standard VSCode helper lookup function, based on the existing search or files libraries. I think this, as when I set `"search.followSymlinks": false"` in my vcscode settings, it uses that flag on the lookup, and so works.

**Comment by matthew-heyner** [comment]
Any updates on this? I'm still experiencing high CPU usage on Copilot version v1.133.0.

**Comment by mbtools** [comment]
Some details and possible solutions:

```env
Version: 1.84.2 (system setup)
Commit: 1a5daa3a0231a0fbba4f14db7ec463cf99d7768e
Date: 2023-11-09T10:51:52.184Z
Electron: 25.9.2
ElectronBuildId: 24603566
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Windows_NT x64 10.0.19045
GitHub Copilot extension: v1.138.0
```

The process triggered on my machine and consuming 100% CPU is

```
"c:\Program Files\Microsoft VS Code\resources\app\node_modules.asar.unpacked\@vscode\ripgrep\bin\rg.exe"
	--files
	--hidden
	--case-sensitive
	--no-require-git
	-g **/.copilotignore
	-g !**/.git
	-g !**/.svn
	-g !**/.hg
	-g !**/CVS
	-g !**/.DS_Store
	-g !**/Thumbs.db
	-g !**/.nyc_output
	-g !**/coverage
	-g !/.idea
	-g !/storage_default_storage
	-g !/.yarn
	--no-ignore
	--follow
	--no-config
	--no-ignore-global
```

There are a few relevant VSCode settings:

`Files: Exclude`

![image](https://github.com/community/community/assets/59966492/5de77047-0a98-4f8a-9692-bc3ec62d9a87)

`Search: Exclude`

![image](https://github.com/community/community/assets/59966492/6a55cf6c-a0a7-48e8-bfd5-66ba49337056)

`Exclude Git Ignore`

![image](https://github.com/community/community/assets/59966492/6e464e43-418b-42db-99f4-4a142e981dc8)

`.vscode/settings.json` (part of the project)

![image](https://github.com/community/community/assets/59966492/d23e18fe-4e30-4c85-8363-48095e320884)

Now you can see how the command parameters are composed. 
 
I did not have `Exclude Git Ignore` on, but activating it didn't influece the search by the Copilot extension. Our `.gitignore` definitely includes `node_modules` but the extension doesn't exclude it as expected. 

Adding `node_modules` to file exclusions (either as setting or in settings.json in project) does work and lets the rg.exe process finish quickly. 

```json
{
  "files.exclude": {
    "**/node_modules": true,
```

However, I believe the Copilot extension should respect the `Exclude Git Ignore` setting in this case (and not make everyone change settings). Thanks!
Marc

**Reply by loren-garcia** [reply]
Totally worked! Thanks!!

**Reply by ljuba-roofr** [reply]
Works for me too, thanks!

**Comment by matthew-heyner** [comment]
Bump

**Comment by dineshdaultani** [comment]
I am having a similar issue in the `Github Copilot: v1.148.0`. Problem is related to the **RAM usage**, I have 64GB RAM available on a remote server and after installing copilot in the remote server/container, gradually the extension is occupying around 10-20GB of RAM memory. After disabling the extension and reloading the workspace, sometimes memory becomes available again, but not always.

**Reply by Konano** [reply]
Same with me

**Comment by jacobshirley** [comment]
Also getting high CPU usage through rg from Copilot. files.exclude works but it hides those files in my editor which I don't really want. Ideally, Copilot would use the .gitignore file as others have suggested

**Comment by OriginalEXE** [comment]
@Akash1134 Any chance you could re-prioritize this one? It seems like it's affecting a good chunk of people, and there is no real workaround that doesn't negatively affect some other functionality.

Thanks!

**Comment by w4ffl35** [comment]
I think @mbtools is correct and came here to post the same. 

1. On Ubuntu the ripgrep process is causing the same issue. 
2. Disabling all extensions fixes the issue. 
3. Enabling Copilot again causes the issue
4. Examine by getting pid and using `ps aux|grep <pid>` - see that nothing in my `.gitignore` file is being respected, see that all settings are set correctly and my project's .gitignore is in the correct place
5. Uninstall VS Code Insiders and reinstalled, issue persists
6. Manually added exclude files as per @mbtools - everything works fine

Started (or became noticeable) today after I updated Ubuntu, but I didn't update VS Code Insiders (was already on the latest as of this writing `code-insiders_1.86.0-1705991583_amd64.deb`)

**Comment by wind57** [comment]
I am quitting on it. I can live without those suggestions, but otherwise I can't do anything in my IDE.

**Comment by pilaoda** [comment]
My temporary solution:
Add `**/node_modules/**/node_modules` in `files.exclude`.
This break the infinite directory loop but still reserve first level node_modules for my need.

**Comment by nsainaney** [comment]
I ran `Help: Start Extension Bisect` and the culprit for me was github.copilot.chat which is a companion extension. Removing this extension solved the high CPU issue. I'm on an M2 Max and the thing drained my battery (which usually lasts about 14hrs) within 2 hrs!

<img width="466" alt="Screenshot 2024-03-13 at 10 12 46 PM" src="https://github.com/community/community/assets/306293/5f375175-8b2c-4a20-9b44-754dbf2c3e7c">

**Reply by mbtools** [reply]
Thanks for adding more proof. I'm at a loss on how to get more attention to this issue 🤷‍♀️

**Comment by illmattic-1** [comment]
I'm having this issue after vscode updated to v1.88. I am using CoPilot v1.178.

This issue only occurs when I have a big project opened. I tried the solution mentioned by @mbtools but it isn't helping. 

```
"files.exclude": {
        "**/.git": true,
        "**/.svn": true,
        "**/.hg": true,
        "**/CVS": true,
        "**/.DS_Store": true,
        "**/Thumbs.db": true,
        "**/node_modules": true,
    },
```

Any help would be appreciated.

**Reply by mbtools** [reply]
Copilot scans all folders that are not excluded. Look whether you have other folders with many files that are not relevant for copilot and exclude them as well.

**Reply by illmattic-1** [reply]
This works but it kind of defeats the purpose because I have to remove so much of my project before it returns to normal that it becomes inconvenient. Before I had the same size project and it wasn't an issue but now when I click the copilot chat, the CPU usage runs up to 80% for VSCode before coming down to a min level of 45% and stays there while my cursor shows that it is continuously loading/buffering. I have a i5-9400 CPU @ 2.90 GHz

**Comment by GregWoods** [comment]
It sounds like a new "copilot-exclude" setting is needed

**Comment by cnicodeme** [comment]
A year later and this is still an issue :/

**Comment by scottw-finao** [comment]
I'm getting a similar issue in PHP storm. copilot sub process is currently consuming 30gb almost 3x what the java process for phpstorm itself is using.

**Comment by cnicodeme** [comment]
That's it, I'm switching to cursor and cancelling my copilot subscription. If they won't fix it, why pay?

https://github.com/settings/billing/summary <= to cancel your subscription to Copilot

**Reply by mbtools** [reply]
Yesterday I thought the performance issue was gone. Then I realized that I was in Cursor 😂

**Comment by christoph-cieslik-sybit** [comment]
Hi everyone, I and my coleages have also a problem using copilot in idea (java projects). The copilot plugin is claiming that much memory that either the tomcat or idea is being killed by the system caused by out of memory  (ubuntu). After disabling the copilot plugin the problem is gone. We also saw, that on September 2 the copilot-language-server process used about 43MB, today it uses >10GB of memory.

**Reply by rocketraman** [reply]
Yes, at the moment on my machine, the `copilot-language-server` is using 482G of virtual memory, and 27G of resident memory.

**Reply by Konano** [reply]
Same here

**Reply by christoph-cieslik-sybit** [reply]
Ok, solution for now: downgraded the copilot plugin version to [1.5.28.7313]. Using the older version works for me for currently. But I cannot say whether it can solve problems for the other of you. 

Download it here and "install plugin from disk" in idea.
https://plugins.jetbrains.com/plugin/17718-github-copilot/versions

**Reply by sureshg** [reply]
On IntelliJ, I had to disable the plugin as it kept consuming excessive memory and killing other processes. `copilot-language-server` consumes `~8GB` and `1000 platform threads` 😳 . After the latest update, the plugin became unusable.

<img width="542" alt="Screenshot 2024-11-22 at 10 31 04 AM" src="https://github.com/user-attachments/assets/eb0edbbd-d5d5-48ae-bc12-71f4596f5dc6">

**Comment by raegen** [comment]
A funny fact is that, it's actually consuming quite a bit of memory in IDEA even when the extension itself is not authorised 
<img width="1440" alt="image" src="https://github.com/user-attachments/assets/9cf9964b-1d1b-412b-8a71-a51ed8a09d10" />

**Comment by ueteanderson** [comment]
I have the same problem in PyCharm. Have tried all the fixes here, but it still uses most of my CPU when the Copilot Chat sidebar is open. Extremely frustrating.

**Reply by CarlossShi** [reply]
For PyCharm, go to `Settings -> Project Structure`. Mark folders containing large files as **Excluded**. This will stop PyCharm indexing large files, and stop Copilot doing so as well. Work like a charm in my project.

**Comment by gezabohus** [comment]
Same problem on Intel Mac, PyCharm. Using 100% of one core.

**Comment by MarkVI2** [comment]
Github copilot really needs a ``copilot.exclude``. Somehow the extension is completely ignoring my ``.gitignore``, ``"files.exlclude" = {...},`` and any other trick to actually stop the recursive directory traversal. If github is incapable to implement the recursive directory traversal, here's my solution: Just form a DAG! This really seems like a stupid bug to leave in the production.

**Comment by mbtools** [comment]
If you read this, please upvote this issue (at the [top](https://github.com/orgs/community/discussions/56755#discussion-5247001)).

**Comment by Hoplite2A** [comment]
This is definitely an issue. My VSCode is taking a substantial amount of resources when GH Copilot is on and even when it is on it doesn't run. IU have to restart the extension in order to be able to add another prompt. 
![GH Copilot Disabled](https://github.com/user-attachments/assets/003e74a9-e289-479d-8c78-d05cdc01b282)
![Gh Copilot Enabled](https://github.com/user-attachments/assets/9c100bc6-da86-4caa-a51b-fa4ba6380353)

**Comment by truong120201** [comment]
It’s 2025 and this is still an issue, I’m canceling it

**Reply by klsbecker** [reply]
Same here, the **rg** process is eating all my CPUs
GitHub Copilot v1.250.0

**Reply by serhiiokhotnik** [reply]
same here 450% cpu

**Reply by mbtools** [reply]
See explanation and possible solution here: 

https://github.com/community/community/discussions/56755#discussioncomment-7634002

**Comment by umithyo** [comment]
![image](https://github.com/user-attachments/assets/19afc582-1357-41ee-b4f2-e636631799c3)

I'm also experiencing this. 
My OS is Ubuntu 24.04
Copilot is installed in VS Code, JetBrains Rider

**Reply by truong120201** [reply]
Hey bro, incase you are working on a nodejs project, try solution above setting.json at github copilot advanced. I tried it and it worked

**Reply by umithyo** [reply]
> Hey bro, incase you are working on a nodejs project, try solution above setting.json at github copilot advanced. I tried it and it worked

Thanks man, but no, I am working on a .NET project at the moment. I'm opening the project on Rider.

**Comment by mbtools** [comment]
To the recent posters, unless your issue is about the `rg` process consuming a very higher amount of CPU, please open a separate issue. Thank you

**Comment by kainjinez** [comment]
<img width="1123" alt="Screenshot 2025-04-01 at 22 15 16" src="https://github.com/user-attachments/assets/d4f6b443-89d2-4ad4-a92f-4acf985e1d09" />
Can't believe it took 32GB RAM. I used this on IntelliJ products like Android Studio and Webstorm.

**Reply by mbtools** [reply]
https://github.com/orgs/community/discussions/56755#discussioncomment-12145059

**Comment by rsebestik** [comment]
This is still an issue? Damn - it takes a lof of CPU and RAM. Please fix this - really thinking going to other plugin...

TIP: i only managed to cool down the CPU, by allowing it to run only on one core via "Set Affinity" in windows task manager for the copilot-language-server.exe:
![image](https://github.com/user-attachments/assets/7d83ca91-8491-4bed-a5cd-9d67b97fcce5)

**Reply by rahulchauhan04** [reply]
I don't understand why this is still happening, coz once I have adapted those changes I am getting coverage of 5-6 gb.of cpu.. but smoothly it Is running...

Simply just tell the copilot to ignore the node modules files...

**Reply by rsebestik** [reply]
> I don't understand why this is still happening, coz once I have adapted those changes I am getting coverage of 5-6 gb.of cpu.. but smoothly it Is running...
> 
> Simply just tell the copilot to ignore the node modules files...

i don't have any node_modules in my pure java project, (nor i have any large files outside of target directories which are marked as excluded...)

**Reply by mbtools** [reply]
https://github.com/orgs/community/discussions/56755#discussioncomment-12145059

**Comment by mieusk** [comment]
I’m experiencing the same issue. GitHub Copilot on the website (github.com/copilot) uses a huge amount of memory, and even after the task is completed, it doesn’t release the memory. Isolated Web Co reach to 8GB of RAM usage, and I need to force stop process because it doesn't release.

**Reply by mbtools** [reply]
https://github.com/orgs/community/discussions/56755#discussioncomment-12145059

**Comment by serhiiokhotnik** [comment]
`rg` issue is still there, narrowed down it's origin to copilot - ran vscode without extensions, then turned off all of them, then turned on one by one - after copilot rg spawned and  took 450% of my MBP M2
